src/common/utxobased/db/newProcessor.ts

```diff
@@ -183,31 +184,178 @@ export async function makeNewProcessor(
+        if (
+          data.lastTouched != null &&
+          data.lastTouched > address.lastTouched
+        ) {
+          address.lastTouched = data.lastTouched
+        }
+
+        // Only update the balance field if one was given and does not equal the existing value
+        if (data.balance != null && data.balance !== address.balance) {
+          address.balance = data.balance
+        }
+
+        // Only update the path field if one was given and currently does not have one
+        // NOTE: Addresses can be stored in the db without a path due to the `EdgeCurrencyEngine.addGapLimitAddresses` function
+        //  Once an address path is known, it should never be updated
+        if (data.path != null && address.path == null) {
+          address.path = data.path
+          await tables.scriptPubkeyByPath.insert(
+            addressPathToPrefix(data.path),
+            data.path.addressIndex,
+            scriptPubkey
+          )
+        }
+
+        // Only update the used flag if one was given and is true
+        if (data.used != null && data.used && !address.used) {
+          address.used = true
+        }
+
+        // check if the lastUsed changed by the update
+        if (address.path != null && address.used) {
+          let [lastUsed] = await tables.lastUsedByFormatPath.query('', [
+            addressPathToPrefix(address.path)
+          ])
+          if (lastUsed == null) lastUsed = -1
+
+          if (lastUsed < address.path.addressIndex) {
+            await tables.lastUsedByFormatPath.insert(
+              '',
+              addressPathToPrefix(address.path),
+              address.path.addressIndex
+            )
+          }
+        }
+
+        await tables.addressByScriptPubkey.insert(
+          '',
+          address.scriptPubkey,
+          address
+        )
+      })
       return
     },
 
-    async fetchAddresses(_args: FetchAddressArgs): Promise<IAddress[]> {
-      return []
+    async fetchAddresses(args: FetchAddressArgs): Promise<IAddress[]> {
+      const { path, scriptPubkey } = args
+      // keep all the addresses to be fetched in this array
+      const addresses: IAddress[] = []
+      // fetch addresses by provided path
+      if (path != null) {
+        await baselets.address(async tables => {
+          const [scriptPubkeyFromPath] = await tables.scriptPubkeyByPath.query(
+            addressPathToPrefix(path),
+            path.addressIndex
+          )
+
+          const [address] = await tables.addressByScriptPubkey.query('', [
+            scriptPubkeyFromPath
+          ])
+
+          addresses.push(address)
+        })
+      }
+      // fetch addresses by provided scriptPubkey
+      if (scriptPubkey != null) {
+        await baselets.address(async tables => {
+          const [address] = await tables.addressByScriptPubkey.query('', [
+            scriptPubkey
+          ])
+          addresses.push(address)
+        })
+      }
```

Comment:
This will only return `[IAddress] | [IAddress, IAddress]`? I wonder why we don't go with `fetchAddress(args: FetchAddressArgs): Promise<IAddress>` and then call twice at the call site; this way the caller knows which address returned is for which query argument (path or scriptPubkey).

Reply:
Yeah, seems cleaner to me as well.